File: shared/src/commonMain/kotlin/com/loca/signal/SignalError.kt
package com.loca.signal

import com.loca.record.FactKind
import kotlin.uuid.Uuid

/** Errors that can occur in the Signal layer. */
sealed class SignalError(message: String) : Exception(message) {
    /** A Fact could not be transformed by the pipeline. */
    class PipelineFailure(val cause: SignalPipelineError) :
        SignalError("Pipeline failure: ${cause.message}")

    /** The underlying signal store failed. */
    class StorageFailure(val underlying: Throwable) :
        SignalError("Storage failure: ${underlying.message}")

    /** A replay operation failed. */
    class ReplayFailure(val reason: String) :
        SignalError("Replay failure: $reason")
}

/** Specific errors from the pipeline transformation step. */
sealed class SignalPipelineError(message: String) : Exception(message) {
    /** The Fact's payload kind does not match its declared FactKind. */
    class PayloadKindMismatch(val factID: Uuid, val factKind: FactKind) :
        SignalPipelineError("Payload does not match kind $factKind for fact $factID")

    /** A required payload field is missing or structurally invalid. */
    class MalformedPayload(val factID: Uuid, val reason: String) :
        SignalPipelineError("Malformed payload for fact $factID: $reason")
}

File: ios/LOCA/V3/Signal/SignalError.swift
import Foundation

/// Errors that can occur in the Signal layer.
enum SignalError: Error, Sendable {
    /// A Fact could not be transformed by the pipeline.
    case pipelineFailure(SignalPipelineError)

    /// The underlying signal store failed.
    case storageFailure(underlying: any Error)

    /// A replay operation failed.
    case replayFailure(message: String)
}

/// Specific errors from the pipeline transformation step.
enum SignalPipelineError: Error, Sendable {
    /// The Fact's payload kind does not match its declared FactKind.
    case payloadKindMismatch(factID: UUID, factKind: FactKind)

    /// A required payload field is missing or structurally invalid.
    case malformedPayload(factID: UUID, reason: String)
}
